src/server/components/charts-engine/components/processor/control-api-context.ts

```diff
@@ -0,0 +1,102 @@
+import type {IChartEditor} from '../../../../../shared';
+import type {ControlShared} from '../../../../modes/charts/plugins/control/types';
+import {resolveIntervalDate, resolveOperation, resolveRelativeDate} from '../utils';
+
+import {getParam} from './paramsUtils';
+import type {ChartApiContext} from './types';
+
+function getOrphanedObject() {
+    return Object.create(null);
+}
+
+type GetControlApiContextArgs = {
+    name: string;
+    params: Record<string, string | string[]>;
+    shared: ControlShared;
+};
+
+export const getControlApiContext = (args: GetControlApiContextArgs): ChartApiContext => {
+    const {name, params, shared = {}} = args;
+
+    const api: IChartEditor = {
+        getSharedData: () => shared,
+        resolveRelative: resolveRelativeDate,
+        resolveInterval: resolveIntervalDate,
+        resolveOperation,
+        getUserLogin: () => '',
+        getUserLang: () => '',
+        getLang: () => '',
+        getTranslation: () => '',
+        getSecrets: () => ({}) as ReturnType<IChartEditor['getSecrets']>,
+        getWidgetConfig: () => undefined,
+        getParams: () => ({}) as ReturnType<IChartEditor['getParams']>,
+        getParam: () => '',
+        getActionParams: () => ({}) as ReturnType<IChartEditor['getActionParams']>,
+        getCurrentPage: () => 0,
+        getSortParams: () => ({}) as ReturnType<IChartEditor['getSortParams']>,
+        updateConfig: () => {},
+        setChartsInsights: () => {},
+        getLoadedData: () => ({}) as ReturnType<IChartEditor['getLoadedData']>,
+        getLoadedDataStats: () => ({}) as ReturnType<IChartEditor['getLoadedDataStats']>,
+        setError: () => {},
+        _setError: () => {},
+        updateHighchartsConfig: () => {},
+        setDataSourceInfo: () => {},
+        setExtra: () => {},
+        updateParams: () => {},
+        updateActionParams: () => {},
+        updateLibraryConfig: () => {},
+        setSideHtml: () => {},
+        setSideMarkdown: () => {},
+        setExportFilename: () => {},
+        wrapFn: () => ({}) as ReturnType<IChartEditor['wrapFn']>,
+        generateHtml: () => ({}) as ReturnType<IChartEditor['generateHtml']>,
+        attachHandler: () => ({}) as ReturnType<IChartEditor['attachHandler']>,
+        attachFormatter: () => ({}) as ReturnType<IChartEditor['attachFormatter']>,
```

Comment:
If a common `IChartEditor` interface is used anyway, then I would like to use a common method - now there is an extra duplication that you can stumble over

Reply:
I've removed it and use chart-api-context